hnswlib/hnswalg.h

```diff
@@ -260,8 +260,8 @@ class HierarchicalNSW : public AlgorithmInterface<dist_t> {
                 tableint candidate_id = *(datal + j);
 //                    if (candidate_id == 0) continue;
 #ifdef USE_SSE
-                _mm_prefetch((char *) (visited_array + *(datal + j + 1)), _MM_HINT_T0);
-                _mm_prefetch(getDataByInternalId(*(datal + j + 1)), _MM_HINT_T0);
+                _mm_prefetch((char *) (visited_array + *(datal + j)), _MM_HINT_T0);
+                _mm_prefetch(getDataByInternalId(*(datal + j)), _MM_HINT_T0);
```

Comment:
I think that fix defeats the purpose of the prefetch, as it need to be done for the next element.
Is there any evidence this prefetch affects the program behavior? 
If so, I guess we probably should come with a different fix, e.g. min(j+1, max_value)

Reply:
Given that a cache line is atleast 64 bytes, does the extra `+1` really matter? What matters is that the loop is moving forward via the `j` variable increments so we keep fetching ahead. We should be able to confirm this via a benchmark: I don't think there will be any degradation in performance.

There are actually a couple of other places the same issue happens, which I just noticed: 

- https://github.com/nmslib/hnswlib/blob/master/hnswlib/hnswalg.h#L346
- https://github.com/nmslib/hnswlib/blob/master/hnswlib/hnswalg.h#L1010

Reply:
Also, I think the proper way to actually use a `_mm_prefetch` is to fetch a full 64-bytes ahead, for example how it's done here by fetching "8 doubles ahead": 

https://stackoverflow.com/questions/73012072/why-such-address-is-used-with-mm-prefetch

Reply:
```C++ 
_mm_prefetch((char *) (visited_array + *(datal + j + 1)), _MM_HINT_T0);
```

+1 fetches the visited bit for the next item, which can be anywhere in the memory compared the previous element. Almost certain different cache and page and at least skylake's cpus were struggling without the prefetches, that is why it ended up in the code despite the obvious issue.

Yeah, it would be great if you could profile the performance on few different x64 platforms and maybe improve the prefetching performance and hopefully fix this out-of-bound issue that made people uneasy. 

I think the proper way to test the performance is to do search on small dimensional data to focus on memory sparsity (I used d=4)  and keep two implementations of the search function (with and without the change), so the performance would not depend on random memory layout changes in the data. That should allow measuring the performance within 1-2% accuracy.